README.md
# neuro

Simple neural network in clojure. 
This code is good for undersanding how neural network work. 
This code in not optimised, and not for using in production.

## Usage

FIXME

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-02-09
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-02-09
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/neuro/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/neuro/compare/0.1.0...0.1.1


test/neuro/core_test.clj
(ns neuro.core-test
  (:require [clojure.test :refer :all]
            [neuro.core :refer :all]
            [neuro.neuron :refer :all]
            [neuro.layer :refer :all]
            [neuro.neuralnetwork :refer :all]
            [neuro.data :refer :all]
            [uncomplicate.neanderthal.math :refer :all]
            [midje.sweet :refer [facts throws => roughly]]))

(def x [1.0 0.86 0.03 0.08 0.903 0.725 0.574 -0.1 -0.203 -0.031 -0.12413793 0.33 0.225])

;; testing neuron

(facts
  "test some neuron functions"
  (same 1) => 1
  (one 5) => 1
  (dot-product [1 2 3] [1 2 3]) => 14
  )

(facts
  "test making input neuron"
  (make-new-input-neuron x) => {:inputs [1.0 0.86 0.03 0.08 0.903 0.725 0.574 -0.1 -0.203 -0.031 -0.12413793 0.33 0.225], :output 4.268862070000001, :weights [1 1 1 1 1 1 1 1 1 1 1 1 1]}
  )

(facts
  "test making neuron"
  (make-neuron x [1 1 1 1 1 1 1 1 1 1 1 1 1] tanh) => {:inputs [1.0 0.86 0.03 0.08 0.903 0.725 0.574 -0.1 -0.203 -0.031 -0.12413793 0.33 0.225], :output 0.9996082054159976, :weights [1 1 1 1 1 1 1 1 1 1 1 1 1]}
  )


(facts
  "test making new neuron"
  (type (make-new-neuron x tanh)) => neuro.neuron.Neuron
  )

;; testing layer

(facts
  "test making input layer"
  (count (map :inputs (make-input-layer x))) => (count x)
  (map :inputs (make-input-layer x)) => (map list x)
  (map :weights (make-input-layer x)) => (map list (replicate (count x) 1))
  (map :output (make-input-layer x)) => (map conj x)
  )

(facts
  "test making hidden layer"
  (count (make-hidden-layer x 32 tanh)) => 32
  (map :inputs (take 1 (make-hidden-layer x 32 tanh))) => (list x)
 )

(facts
  "test making output layer"
  (:inputs (make-output-layer x tanh)) => x
  (<= (:output (make-output-layer x tanh)) 1)  => true
  (>= (:output (make-output-layer x tanh)) 0)  => true
  )

(facts
  "test generate layer"
  (generate-layer (vec (map vector x)) (vec (replicate (count x) [1])) same) => (make-input-layer x)
  )

;; testing network

(facts
  "test network"
  (count (network x 24)) => 3
  (vec (map :inputs (-> (:input-layer (network x 24))))) => (map list x)
  (count (:hidden-layer (network x 24))) => 24
  (map :inputs (take 1 (:hidden-layer (network x 24)))) => (vector (map conj x))
  (<= (:output (:output-layer (network x 24))) 1) => true
  (>= (:output (:output-layer (network x 24))) 0) => true
  )







src/neuro/data.clj
(ns neuro.data
  (:require [clojure.string :as string]))


(defn parse-float [s]
  (Float/parseFloat s))

(defn read-data-training
  "Read the csv file, split out each line and then each number, parse the tokens and break up the numbers so that the last is the target and everything else is the feature vector."
  []
  (as-> (slurp "resources/training_norm_podaci.csv") d
        (string/split d #"\n")
        (map #(string/split % #",") d)
        (map #(map parse-float %) d)
        (map (fn [s] {:x  (vec (drop-last s))  :y (last s)}) d)))

(defn read-data-test
  "Read the csv file, split out each line and then each number, parse the tokens and break up the numbers so that the last is the target and everything else is the feature vector."
  []
  (as-> (slurp "resources/test_norm_podaci.csv") d
        (string/split d #"\n")
        (map #(string/split % #",") d)
        (map #(map parse-float %) d)
        (map (fn [s] {:x  (vec (drop-last s))  :y (last s)}) d)))


src/neuro/layer.clj
(ns neuro.layer
  (:require [neuro.neuron :refer :all]))

(defn make-input-layer
  "make a new input layer with input vector x"
  [x]
  (map #(make-new-input-neuron (vector %)) x))

(defn make-hidden-layer
  "make a new hidden layer with random weights and input vector x"
  ([x num-neurons o-func]
   (map #(make-new-neuron % o-func) (replicate num-neurons x))))

(defn make-output-layer
  "make a new output layer with random weight and input vector x"
  [x o-func]
  (make-new-neuron x o-func))

(defn generate-layer
  "make a new layer with input and weight vectors"
  [x w o-func]
  (for [[x w] (map list x w)]
    (make-neuron x w o-func)))

(defn change-weights-value
  "only change weight vector on neural network"
  [n weigths o-func]
  (for [[x w] (map list (map :inputs (-> n)) weigths) ]
     (make-neuron x w o-func)))

(defn change-neuron-weights-value
  "make a output layer with a one neuron with input and weight vectors"
  [inputs weigths o-func]
    (make-neuron inputs weigths o-func))







src/neuro/neuron.clj
(ns neuro.neuron)

;; neuron

(defrecord Neuron [inputs           ;; input vector
                   weights          ;; weight vector
                   output           ;; output value
                   ])

(defn same
  "f(x)=x"
  [x]
  (-> x))

(defn one
  "f(x)=1"
  [x]
  (-> 1))

(defn dot-product
  "Dot-product - idiomatic clojure"
  [xs ys]
  (reduce + (map * xs ys)))

(defn output
  "output from neuron"
  [x w o-func]
  (o-func (dot-product x w)))

(defn make-neuron
  "make a neuron from vectors"
  [x w o-func]
  (->Neuron x w (output x w o-func)))

(defn random-number
  "random number in interval [0 .. 0.1]"
  [x]
  (rand 0.1))

(defn make-new-neuron
  "make a neuron from vectors with random weights"
  [x o-func]
  (make-neuron x (vec (map random-number x)) o-func))

(defn make-new-input-neuron
  "make a new input neuron, input=output"
  [x]
  (make-neuron x (vec (map one x)) same))



src/neuro/neuralnetwork.clj
(ns neuro.neuralnetwork
  (:require [neuro.neuron :refer :all]
            [neuro.layer :refer :all]
            [neuro.data :refer :all]
            [uncomplicate.neanderthal.core :refer :all]
            [uncomplicate.neanderthal.math :refer :all]
            [uncomplicate.neanderthal.native :refer :all]))


(defrecord Neuronetwork [
                   input-layer           ;; input set of Neurons
                   hidden-layer          ;; hidden set of Neurons
                   output-layer          ;; output set of Neurons
                   ])

(defn network
  "make a new neural network with input vector and desired number of hidden neurons"
  [x dim-hidden-layers]
  (let [
        input-layer (make-input-layer x)
        hidden-layer1 (make-hidden-layer (vec (map :output (-> input-layer))) dim-hidden-layers tanh)
        output-layer (make-output-layer (vec (map :output (-> hidden-layer1))) tanh)
        ]
    (->Neuronetwork input-layer
                    hidden-layer1
                    output-layer)))

(defn dtanh
  "help function"
  [y]
  (tanh (- 1 (sqr y))))

(defn output-deltas2
  "delta function for output layer"
  [out target]
  (* (dtanh out) (- target out)))

(defn hidden-error
  "help function for delta function for hidden layer"
  [o-delta weights]
  (dot-product weights (vec (replicate (count weights) o-delta))))

(defn hidden-deltas
  "delta function for hidden layer"
  [o-delta weights output-hidden-layer]
  (for [[w] (map list weights)]
    (let [error (hidden-error o-delta w)]
      (vec (map #(* o-delta %) (vec (map dtanh output-hidden-layer)))))))

(defn change-output-weights-vec
  "make new vector for change output weights"
  [o-delta learning-speed hidden-outputs-vec]
  (vec (map #(* (* o-delta learning-speed) %) hidden-outputs-vec)))

(defn change-hidden-weight-vec
  "make new vector for change hidden weights"
  [h-deltas learning-speed input-hidden-layer]
  (vec (for [[h i] (map list h-deltas input-hidden-layer)]
         (vec (for [[hh ii] (map list h i)]
                (* (* hh learning-speed) ii))))))

(defn changed-weights-vec
  "calculate new weight vector for output layer"
  [old-weights change-vec]
  (vec (map + old-weights change-vec)))

(defn changed-hidden-weights-vec
  "calculate new weight vector for hidden layer"
  [old-weights change-vec]
  (vec (for [[o c] (map list old-weights change-vec)]
         (vec (map + o c)))))

(defn change-input-vec-nn
  "change input vector on neural network and feed forward propagation"
  [nn x]
  (let [prepare-input (vec (map vector x))
        input-layer (generate-layer prepare-input (vec (replicate (count x) [1])) same)
        hidden-layer (generate-layer (vec (replicate (count (:hidden-layer nn)) x))  (vec (map :weights (:hidden-layer nn))) tanh)
        output-layer (make-neuron (vec (map :output hidden-layer)) (vec (:weights (:output-layer nn))) tanh)]
    (->Neuronetwork input-layer
                    hidden-layer
                    output-layer)))

(defn back-propagation
  "one iteration of backpropagation for neural network with desired params"
  [nn target speed-learning]
  (let [input-layer (:input-layer nn)
        hidden-layer (generate-layer
                       (vec (map :inputs (-> (:hidden-layer nn))))
                       (changed-hidden-weights-vec
                         (vec (map :weights (-> (:hidden-layer nn))))
                         (change-hidden-weight-vec
                           (hidden-deltas (output-deltas2 (:output (:output-layer nn)) target)
                                          (vec (map :weights (-> (:hidden-layer nn))))
                                          (vec (map :output (-> (:hidden-layer nn))))
                                          )
                           speed-learning
                           (vec (map :inputs (-> (:hidden-layer nn))))))tanh)
        output-layer (change-neuron-weights-value (vec (map :output hidden-layer))
                                                  (changed-weights-vec (:weights (:output-layer nn)) (change-output-weights-vec (output-deltas2 (:output (:output-layer nn)) target) speed-learning  (vec (map :output hidden-layer))))
                                                  tanh)]
    (->Neuronetwork input-layer
                    hidden-layer
                    output-layer)))

(defn learn
  "learning neural network for one target value"
  [nn current-number max-number speed-learning target]
  (let [num (+ current-number 1)
        ]
    (if (< num max-number)
      (learn (back-propagation nn target speed-learning) num max-number speed-learning target)
      (back-propagation nn target speed-learning))))

(defn train-network
  "train of neural network for all data"
  [atom-nn num-per-line speed-learning]
  (let [input-vec (vec (map :x (read-data-training)))
        target-vec (map :y (read-data-training))]
    (for [[x y] (map list input-vec target-vec)]
      (do
        (reset! atom-nn (learn (change-input-vec-nn @atom-nn x) 0 num-per-line speed-learning y))
        ))))

(defn evaluation
  "evaluation neural network - detail report"
  [atom-nn]
  (let [input-vec (vec (map :x (read-data-test)))
        target-vec (map :y (read-data-test))]
    (for [[x y] (map list input-vec target-vec)]
      {:output (:output (:output-layer (change-input-vec-nn @atom-nn x))) :target y
       :percent    (* 100 (- (/ (:output (:output-layer (change-input-vec-nn @atom-nn x))) y) 1))
       :percent-abs (abs (* 100 (- (/ (:output (:output-layer (change-input-vec-nn @atom-nn x))) y) 1)))}
      )))

(defn evaluation_sum
  "evaluation neural network - average report by deviations"
  [atom-nn]
  (let [u (count (map :percent (evaluation atom-nn)))
        s (reduce + (map :percent (evaluation atom-nn)))
        ]
    (/ s u)))

(defn evaluation_sum_abs
  "evaluation neural network - average report by absolute deviations"
  [atom-nn]
  (let [u (count (map :percent-abs (evaluation atom-nn)))
        s (reduce + (map :percent-abs (evaluation atom-nn)))
        ]
    (/ s u)))





















src/neuro/example.clj
(ns neuro.example
  (:require [neuro.neuron :refer :all]
            [neuro.layer :refer :all]
            [neuro.neuralnetwork :refer :all]
            [neuro.data :refer :all]
            [uncomplicate.neanderthal.core :refer :all]
            [uncomplicate.neanderthal.math :refer :all]
            [uncomplicate.neanderthal.native :refer :all]
            [criterium.core :refer :all]))


(def x [1.0 0.86 0.03 0.08 0.903 0.725 0.574 -0.1 -0.203 -0.031 -0.12413793 0.33 0.225])
(def xx [0 0.29	0.13 0.08	0.868	0.859	0.618	-0.208 -0.319 -0.125 -0.2	0.96 0.575])
(def xxx [1	1	1	1	0.918	0.87 0.615 -0.056	-0.144 0.047 -0.004344828	0.25 0.2])


(def test-network (network x 24))
(def vladatt (atom test-network))


(-> test-network)

(:input-layer test-network)
(:hidden-layer test-network)
(:output-layer test-network)
(:inputs (:output-layer test-network))
(:weights (:output-layer test-network))
(:output (:output-layer test-network))
(vec (map :weights (-> (:hidden-layer test-network))))


(for [a (replicate 10 1)]
  (train-network vladatt a 0.05))

(str (for [a (replicate 10 1)]
       (train-network vladatt a 0.05)
       ))

(while (< 2.8 (evaluation_sum_abs vladatt))
  (str
    (for [a (replicate 20 1)]
      (train-network vladatt a 0.05)
      )))

(reduce max (map :percent-abs (evaluation vladatt)))
(:output (:output-layer (change-input-vec-nn @vladatt [0	0.71	0.94	1	0.897	0.816	0.612	0.003	-0.05	0.094	0.134827586	0.42	0.375])))



